src/compiler/checker.ts

```diff
@@ -43614,6 +43614,9 @@ namespace ts {
         }
 
         function checkGrammarClassLikeDeclaration(node: ClassLikeDeclaration): boolean {
+            if (node.parent && isIfStatement(node.parent)) {

```

Comment:
Is this the right fix? I'm also wondering about cases like:

```ts
declare const condition: boolean;

if (condition) console.log(); else class X {}

function foo() {
	return class X {};
}
```

Which all should error (and maybe more). Maybe it's that the check should be checking if the parent is some sort of block?

Reply:
Also, I'm a little confused as to why this is in the checker; are these grammar errors that should be in the parser?

Reply:
Some grammar errors are computed in the check phase so that the parser can have better error recovery

Reply:
e.g. in this case, it's "better" for the parser to just accept the invalid class decl and then flag it later during a subsequent walk, as opposed to issuing individual errors for ~every token of the class

Reply:
I see, I was just guessing that it'd be like the other errors generated in the binder, e.g. the variable redeclaration stuff that was recently moved there to allow the errors to show in JS.

Reply:
> e.g. in this case, it's "better" for the parser to just accept the invalid class decl and then flag it later during a subsequent walk, as opposed to issuing individual errors for ~every token of the class

It'd do that? I would have guessed that it'd recover by parsing the class but emitting an error anyway along with the fully parsed class node.

Reply:
Thanks for the review! I will implement also the other checks, should I move the checks or let them in the `checker`?

Reply:
You should be able to just leave them in the `checkGrammar` functions (and assume my intuition here is wrong).

Reply:
Ok thanks, then I will add other checks there to also avoid cases that you pointed out :)

Reply:
I sort of misspoke, it'd still be in the function you just edited because that's the one that's checking a class declaration, it's just that you need to check for other wrong parent node types. Or, the other way around in that maybe the fix is to error if the parent is not a block or something.

Reply:
Thanks for the clarification!

I was also wondering on something like this:
```
while (class C {}) {
    break;
}
```
this should also be included right?

Also I noticed that all the cases we are mentioning, for example:
```
function foo() {
    console.log(12);
	return class X {};
}
const A = foo()
```
Is actually possible now and works in js, so this would be a breaking change, right?

Reply:
Probably, if it doesn't have an error already. This supports my "parent is not a block" rule.

For the second case, I think I tested that and it failed. Are you seeing it work?

Reply:
The last test seems  to work now, I checked also with the [last release](https://www.typescriptlang.org/play?ts=4.6.2#code/GYVwdgxgLglg9mABMOcAUBKRBvAUIgxCBAZzgBsBTAOnLgHM0BGAJgwG5cBIAJ0qhA8kEcgEMSJRAA0cAX06zcxMCSiIAgogC8yVJiA) in the playground and with the [nightly build](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220406#code/GYVwdgxgLglg9mABMOcAUBKRBvAUIgxCBAZzgBsBTAOnLgHM0BGAJgwG5cBIAJ0qhA8kEcgEMSJRAA0cAX06zcxMCSiIAgogC8yVJiA)

Reply:
The parent is a block rule makes absolute sense yes, thanks!

Reply:
I implemented the fix with the block check :)

Reply:
Cool, thanks!

Reply:
mmm I am seeing this failing some compiler tests (sorry for this, my own computer is really to slow so I usually push to use the github infrastructure, I have tried the one test I modified but not the rest)